DOCS.md

```diff
@@ -0,0 +1,246 @@
+
+### Example
+
+Typically a recursive function with two 'child tasks' is parallelized as follows:
+
+```c
+TASK_1(int, fibonacci, int, n)   // macro to create Lace functions (can be in header file)
+
+int fibonacci(LaceWorker* lw, int n) {
+    if(n < 2) return n;
+    fibonacci_SPAWN(lw, n-1);    // SPAWN a task (fork)
+    int a = fibonacci(lw, n-2);  // run another task in parallel
+    int b = fibonacci_SYNC(lw);  // SYNC the spawned task (join)
+    return a + b;
+}
+
+int main(int argc, char** argv)
+{
+    int n_workers = 4;           // create 4 workers
+                                 // use 0 to automatically use all available cores
+    int dqsize = 0;              // use default task deque size
+    int stacksize = 0;           // use default program stack size
+
+    lace_start(n_workers, dqsize, stacksize);
+    int result = fibonacci_RUN(42);
+    printf("fibonacci(42) = %d\n", result);
+    lace_stop();
+}
+```
+
+See further the [benchmarks](benchmarks/) directory for more examples.
+
+### Interrupting
+
+Lace offers two methods to interrupt currently running tasks and run something else:
+- the `fib_NEWFRAME(40)` function halts current tasks and offers the `fib(40)` task to the framework.
+- the `fub_TOGETHER(40)` function halts current tasks and lets **all Lace workers** execute a copy of the given `fib(40)` task.
+
+The `TOGETHER` functions are typically used to initialize thread-local variables on each worker, or to perform some other operation on each individual worker thread. The `NEWFRAME` functions are used for example to implement stop-the-world garbage collection, interrupting current tasks in order to run a more important task first.
+
+Interrupting is cooperative. Lace checks for interrupting tasks when stealing work, i.e., during `SYNC` or when idle.
+Long-running tasks should use `lace_check_yield` to manually check if an interruption is occuring.
+
+---
+
+## Lifecycle
+
+These methods control the Lace workers, starting and stopping the workers, and suspending and resuming them. These methods must be called from outside the Lace framework, i.e., not from a Lace worker thread.
+
+A typical Lace program starts with `lace_start` and ends with `lace_stop`, and uses `<task>_RUN(...)` to let the Lace worker threads execute tasks.
+
+When Lace is started, worker threads are created and immediately start busy-waiting for work.
+Each thread allocates its own task queue of the requested size.
+
+If `LACE_USE_MMAP` is set, then the queue is preallocated in virtual memory, and real memory is allocated by the OS on demand.
+If `LACE_USE_HWLOC` is set, the worker threads will pin to a CPU core.
+
+Lace workers busy-wait for tasks to steal, increasing the CPU load to 100%.
+Use `lace_suspend` and `lace_resume` (in a non-Lace thread) to temporarily stop the work-stealing framework.
+Suspending and resuming typically requires less than 1 ms.
+
+
+
+
```

Comment:
Uneeded vertical whitespace (a problem when reading the file with `cat`). But, speaking of `cat`: you may want to fill this and other `*.md` files to a column width of 79.